src/promise.ts

```diff
@@ -0,0 +1,53 @@
+/**
+ * A deferred Promise.
+ *
+ * A deferred Promise is one that can be resolved or rejected independently of
+ * the Promise construction.
+ */
+export type DeferredPromise = {
+  /**
+   * The Promise that has been deferred.
+   */
+  promise: Promise<void>;
+  /**
+   * A function that resolves the Promise.
+   */
+  resolve: () => void;
```

Comment:
We also use deferred promises that return a result, not just `Promise<void>`

Reply:
Good point, I had wondered about this. I've mostly seen this with no response, but I guess we may as well make this generic. I'll update this to allow different return values.

Reply:
Added type parameter plus a new test here: https://github.com/MetaMask/utils/pull/164/commits/fc7715e8814eb15556d854d686070654d65b8bab